lib/membrane_h264_plugin/parser/nalu_parser.ex

```diff
@@ -3,42 +3,65 @@ defmodule Membrane.H264.Parser.NALuParser do
   A module providing functionality of parsing a stream of binaries, out of which each
   is a payload of a single NAL unit.
   """
+
+  alias Membrane.H264.Parser
   alias Membrane.H264.Parser.{NALu, NALuTypes}
-  alias Membrane.H264.Parser.NALuParser.SchemeParser
-  alias Membrane.H264.Parser.NALuParser.Schemes
+  alias Membrane.H264.Parser.NALuParser.{SchemeParser, Schemes}
+
+  @annexb_prefix_code <<0, 0, 0, 1>>
 
   @typedoc """
   A structure holding the state of the NALu parser.
   """
   @opaque t :: %__MODULE__{
-            scheme_parser_state: SchemeParser.t()
+            scheme_parser_state: SchemeParser.t(),
+            input_stream_structure: Parser.stream_structure(),
+            output_stream_structure: Parser.stream_structure(),
+            stable_reprefixing?: boolean()
           }
-
-  defstruct scheme_parser_state: SchemeParser.new()
+  @enforce_keys [:input_stream_structure, :output_stream_structure]
+  defstruct @enforce_keys ++
+              [scheme_parser_state: SchemeParser.new(), stable_reprefixing?: true]
 
   @doc """
   Returns a structure holding a clear NALu parser state.
   """
-  @spec new() :: t()
-  def new(), do: %__MODULE__{}
+  @spec new(
+          Parser.stream_structure(),
+          Parser.stream_structure(),
+          boolean()
+        ) :: t()
+  def new(
+        input_stream_structure \\ :annexb,
+        output_stream_structure \\ :annexb,
+        stable_reprefixing? \\ true
+      ) do
+    %__MODULE__{
+      input_stream_structure: input_stream_structure,
+      output_stream_structure: output_stream_structure,
+      stable_reprefixing?: stable_reprefixing?
+    }
+  end
 
   @doc """
   Parses a binary representing a single NALu.
 
   Returns a structure that
   contains parsed fields fetched from that NALu.
-  The input binary is expected to contain the prefix, defined as in
-  the *"Annex B"* of the *"ITU-T Rec. H.264 (01/2012)"*.
+  The input binary is expected to contain one of:
+  * prefix defined as the *"Annex B"* of the *"ITU-T Rec. H.264 (01/2012)"*.
+  * prefix of size defined in state describing the length of the NALU in bytes, as described in *ISO/IEC 14496-15*.
   """
-  @spec parse(binary(), t()) :: {NALu.t(), t()}
-  def parse(nalu_payload, state) do
-    {prefix_length, nalu_payload_without_prefix} =
-      case nalu_payload do
-        <<0, 0, 1, rest::binary>> -> {3, rest}
-        <<0, 0, 0, 1, rest::binary>> -> {4, rest}
+  @spec parse(binary(), t(), boolean()) :: {NALu.t(), t()}
+  def parse(nalu_payload, state, payload_prefixed? \\ true) do
```

Comment:
I am not sure if we should change the stream format (delimeter or lenght-prefix) of the payload here. I would add a separate function for that purpose and let the `parse` return a sort of a "common" representation (without delimeter nor prefix).

Reply:
although upon further pondering I think having payload unprefixed inside NALu.t() structs is worth considering, we could prefix them just before buffering them and having them unprefixed in code would spare us some unprefixing